### ISSUE TITLE
chore: netlify-cli package install-size analysis

### ISSUE THREAD

**Issue by iChenLei** [netlify/cli#3302]
### TL;DR
Too many big-size binary npm package in dependencies list, so `netlify-cli`'s install size is very huge.

### List all  netlify-cli(v6.8.6) deps
| package | install-size | semver |
| :------- | :------- | :----: |
| @netlify/build | [![install size](https://packagephobia.com/badge?p=@netlify/build)](https://packagephobia.com/result?p=@netlify/build) | ^18.7.4 |
| @netlify/config | [![install size](https://packagephobia.com/badge?p=@netlify/config)](https://packagephobia.com/result?p=@netlify/config) | ^15.5.1 |
| @netlify/framework-info | [![install size](https://packagephobia.com/badge?p=@netlify/framework-info)](https://packagephobia.com/result?p=@netlify/framework-info) | ^5.9.1 |
| @netlify/local-functions-proxy | [![install size](https://packagephobia.com/badge?p=@netlify/local-functions-proxy)](https://packagephobia.com/result?p=@netlify/local-functions-proxy) | ^1.1.1 |
| @netlify/plugin-edge-handlers | [![install size](https://packagephobia.com/badge?p=@netlify/plugin-edge-handlers)](https://packagephobia.com/result?p=@netlify/plugin-edge-handlers) | ^1.11.22 |
| @netlify/plugins-list | [![install size](https://packagephobia.com/badge?p=@netlify/plugins-list)](https://packagephobia.com/result?p=@netlify/plugins-list) | ^3.6.0 |
| @netlify/routing-local-proxy | [![install size](https://packagephobia.com/badge?p=@netlify/routing-local-proxy)](https://packagephobia.com/result?p=@netlify/routing-local-proxy) | ^0.31.0 |
| @netlify/zip-it-and-ship-it | [![install size](https://packagephobia.com/badge?p=@netlify/zip-it-and-ship-it)](https://packagephobia.com/result?p=@netlify/zip-it-and-ship-it) | 4.20.0 |
| @oclif/command | [![install size](https://packagephobia.com/badge?p=@oclif/command)](https://packagephobia.com/result?p=@oclif/command) | ^1.6.1 |
| @oclif/config | [![install size](https://packagephobia.com/badge?p=@oclif/config)](https://packagephobia.com/result?p=@oclif/config) | ^1.15.1 |
| @oclif/errors | [![install size](https://packagephobia.com/badge?p=@oclif/errors)](https://packagephobia.com/result?p=@oclif/errors) | ^1.3.4 |
| @oclif/parser | [![install size](https://packagephobia.com/badge?p=@oclif/parser)](https://packagephobia.com/result?p=@oclif/parser) | ^3.8.4 |
| @oclif/plugin-help | [![install size](https://packagephobia.com/badge?p=@oclif/plugin-help)](https://packagephobia.com/result?p=@oclif/plugin-help) | ^3.0.0 | 
| @oclif/plugin-not-found | [![install size](https://packagephobia.com/badge?p=@oclif/plugin-not-found)](https://packagephobia.com/result?p=@oclif/plugin-not-found) | ^1.1.4 |
| @oclif/plugin-plugins | [![install size](https://packagephobia.com/badge?p=@oclif/plugin-plugins)](https://packagephobia.com/result?p=@oclif/plugin-plugins) | ^1.9.3 |
| @octokit/rest | [![install size](https://packagephobia.com/badge?p=@octokit/rest)](https://packagephobia.com/result?p=@octokit/rest) | ^18.0.0 |
| @sindresorhus/slugify | [![install size](https://packagephobia.com/badge?p=@sindresorhus/slugify)](https://packagephobia.com/result?p=@sindresorhus/slugify) | ^1.1.0 |
| @ungap/from-entries | [![install size](https://packagephobia.com/badge?p=@ungap/from-entries)](https://packagephobia.com/result?p=@ungap/from-entries) | ^0.2.1 |
| ansi-styles | [![install size](https://packagephobia.com/badge?p=ansi-styles)](https://packagephobia.com/result?p=ansi-styles) | ^5.0.0 |
| ascii-table | [![install size](https://packagephobia.com/badge?p=ascii-table)](https://packagephobia.com/result?p=ascii-table) | 0.0.9 | 
| backoff | [![install size](https://packagephobia.com/badge?p=backoff)](https://packagephobia.com/result?p=backoff) | ^2.5.0 |
| better-opn | [![install size](https://packagephobia.com/badge?p=better-opn)](https://packagephobia.com/result?p=better-opn) | ^2.1.1 |
| body-parser | [![install size](https://packagephobia.com/badge?p=body-parser)](https://packagephobia.com/result?p=body-parser) | ^1.19.0 |
| boxen | [![install size](https://packagephobia.com/badge?p=boxen)](https://packagephobia.com/result?p=boxen) | ^5.0.0 |
| chalk | [![install size](https://packagephobia.com/badge?p=chalk)](https://packagephobia.com/result?p=chalk) | ^4.0.0 |
| chokidar | [![install size](https://packagephobia.com/badge?p=chokidar)](https://packagephobia.com/result?p=chokidar) | ^3.0.2 |
| ci-info | [![install size](https://packagephobia.com/badge?p=ci-info)](https://packagephobia.com/result?p=ci-info) | ^3.0.0 |
| clean-deep | [![install size](https://packagephobia.com/badge?p=clean-deep)](https://packagephobia.com/result?p=clean-deep) | ^3.0.2 | 
| cli-ux | [![install size](https://packagephobia.com/badge?p=cli-ux)](https://packagephobia.com/result?p=cli-ux) |^5.5.1 |
| concordance |[![install size](https://packagephobia.com/badge?p=concordance)](https://packagephobia.com/result?p=concordance) | ^5.0.0 |
| configstore | [![install size](https://packagephobia.com/badge?p=configstore)](https://packagephobia.com/result?p=configstore) | ^5.0.0 |
| content-type | [![install size](https://packagephobia.com/badge?p=content-type)](https://packagephobia.com/result?p=content-type) | ^1.0.4 |
| cookie | [![install size](https://packagephobia.com/badge?p=cookie)](https://packagephobia.com/result?p=cookie) | ^0.4.0 |
| copy-template-dir | [![install size](https://packagephobia.com/badge?p=copy-template-dir)](https://packagephobia.com/result?p=copy-template-dir) | ^1.4.0 |
| debug | [![install size](https://packagephobia.com/badge?p=debug)](https://packagephobia.com/result?p=debug) | ^4.1.1 |
| decache | [![install size](https://packagephobia.com/badge?p=decache)](https://packagephobia.com/result?p=decache) | ^4.6.0 |
| del | [![install size](https://packagephobia.com/badge?p=del)](https://packagephobia.com/result?p=del) | ^6.0.0 |
| dot-prop | [![install size](https://packagephobia.com/badge?p=dot-prop)](https://packagephobia.com/result?p=dot-prop) | ^6.0.0 |
| dotenv | [![install size](https://packagephobia.com/badge?p=dotenv)](https://packagephobia.com/result?p=dotenv) | ^10.0.0|
| env-paths | [![install size](https://packagephobia.com/badge?p=env-paths)](https://packagephobia.com/result?p=env-paths) | ^2.2.0 |
|envinfo|[![install size](https://packagephobia.com/badge?p=envinfo)](https://packagephobia.com/result?p=envinfo)|^7.3.1|
|execa|[![install size](https://packagephobia.com/badge?p=execa)](https://packagephobia.com/result?p=execa)|^5.0.0|
|express|[![install size](https://packagephobia.com/badge?p=express)](https://packagephobia.com/result?p=express)|^4.17.1|
|express-logging|[![install size](https://packagephobia.com/badge?p=express-logging)](https://packagephobia.com/result?p=express-logging)|^1.1.1|
|filter-obj|[![install size](https://packagephobia.com/badge?p=filter-obj)](https://packagephobia.com/result?p=filter-obj)|^2.0.1|
|find-up|[![install size](https://packagephobia.com/badge?p=find-up)](https://packagephobia.com/result?p=find-up)|^5.0.0|
|flush-write-stream|[![install size](https://packagephobia.com/badge?p=flush-write-stream)](https://packagephobia.com/result?p=flush-write-stream)|^2.0.0|
|folder-walker|[![install size](https://packagephobia.com/badge?p=folder-walker)](https://packagephobia.com/result?p=folder-walker)|^3.2.0|
|from2-array|[![install size](https://packagephobia.com/badge?p=from2-array)](https://packagephobia.com/result?p=from2-array)|^0.0.4|
|fuzzy|[![install size](https://packagephobia.com/badge?p=fuzzy)](https://packagephobia.com/result?p=fuzzy)|^0.1.3|
|get-port|[![install size](https://packagephobia.com/badge?p=get-port)](https://packagephobia.com/result?p=get-port)|^5.1.0|
|gh-release-fetch|[![install size](https://packagephobia.com/badge?p=gh-release-fetch)](https://packagephobia.com/result?p=gh-release-fetch)|^2.0.0|
|git-repo-info|[![install size](https://packagephobia.com/badge?p=git-repo-info)](https://packagephobia.com/result?p=git-repo-info)|^2.1.0|
|gitconfiglocal|[![install size](https://packagephobia.com/badge?p=gitconfiglocal)](https://packagephobia.com/result?p=gitconfiglocal)|^2.1.0|
|hasbin|[![install size](https://packagephobia.com/badge?p=hasbin)](https://packagephobia.com/result?p=hasbin)|^1.2.3|
|hasha|[![install size](https://packagephobia.com/badge?p=hasha)](https://packagephobia.com/result?p=hasha)|^5.2.2|
|http-proxy|[![install size](https://packagephobia.com/badge?p=http-proxy)](https://packagephobia.com/result?p=http-proxy)|^1.18.0|
|http-proxy-middleware|[![install size](https://packagephobia.com/badge?p=http-proxy-middleware)](https://packagephobia.com/result?p=http-proxy-middleware)|^1.0.0|
|https-proxy-agent|[![install size](https://packagephobia.com/badge?p=https-proxy-agent)](https://packagephobia.com/result?p=https-proxy-agent)|^5.0.0|
|inquirer|[![install size](https://packagephobia.com/badge?p=inquirer)](https://packagephobia.com/result?p=inquirer)|^6.5.1|
|inquirer-autocomplete-prompt|[![install size](https://packagephobia.com/badge?p=inquirer-autocomplete-prompt)](https://packagephobia.com/result?p=inquirer-autocomplete-prompt)|^1.0.1|
|is-docker|[![install size](https://packagephobia.com/badge?p=is-docker)](https://packagephobia.com/result?p=is-docker)|^2.0.0|
|is-plain-obj|[![install size](https://packagephobia.com/badge?p=is-plain-obj)](https://packagephobia.com/result?p=is-plain-obj)|^3.0.0|
|isexe|[![install size](https://packagephobia.com/badge?p=isexe)](https://packagephobia.com/result?p=isexe)|^2.0.0|
|jwt-decode|[![install size](https://packagephobia.com/badge?p=jwt-decode)](https://packagephobia.com/result?p=jwt-decode)|^3.0.0|
|lambda-local|[![install size](https://packagephobia.com/badge?p=lambda-local)](https://packagephobia.com/result?p=lambda-local)|^2.0.0|
|listr|[![install size](https://packagephobia.com/badge?p=listr)](https://packagephobia.com/result?p=listr)|^0.14.3|
|locate-path|[![install size](https://packagephobia.com/badge?p=locate-path)](https://packagephobia.com/result?p=locate-path)|^6.0.0|
|lodash|[![install size](https://packagephobia.com/badge?p=lodash)](https://packagephobia.com/result?p=lodash)|^4.17.20|
|log-symbols|[![install size](https://packagephobia.com/badge?p=log-symbols)](https://packagephobia.com/result?p=log-symbols)|^4.0.0|
|make-dir|[![install size](https://packagephobia.com/badge?p=make-dir)](https://packagephobia.com/result?p=make-dir)|^3.0.0|
|memoize-one|[![install size](https://packagephobia.com/badge?p=memoize-one)](https://packagephobia.com/result?p=memoize-one)|^5.2.1|
|minimist|[![install size](https://packagephobia.com/badge?p=minimist)](https://packagephobia.com/result?p=minimist)|^1.2.5|
|multiparty|[![install size](https://packagephobia.com/badge?p=multiparty)](https://packagephobia.com/result?p=multiparty)|^4.2.1|
|netlify|[![install size](https://packagephobia.com/badge?p=netlify)](https://packagephobia.com/result?p=netlify)|^8.0.0|
|netlify-headers-parser|[![install size](https://packagephobia.com/badge?p=netlify-headers-parser)](https://packagephobia.com/result?p=netlify-headers-parser)|^4.0.1|
|netlify-redirect-parser|[![install size](https://packagephobia.com/badge?p=netlify-redirect-parser)](https://packagephobia.com/result?p=netlify-redirect-parser)|^11.0.2|
|netlify-redirector|[![install size](https://packagephobia.com/badge?p=netlify-redirector)](https://packagephobia.com/result?p=netlify-redirector)|^0.2.1|
|node-fetch|[![install size](https://packagephobia.com/badge?p=node-fetch)](https://packagephobia.com/result?p=node-fetch)|^2.6.0|
|node-version-alias|[![install size](https://packagephobia.com/badge?p=node-version-alias)](https://packagephobia.com/result?p=node-version-alias)|^1.0.1|
|oclif-plugin-completion|[![install size](https://packagephobia.com/badge?p=oclif-plugin-completion)](https://packagephobia.com/result?p=oclif-plugin-completion)|^0.6.0|
|omit.js|[![install size](https://packagephobia.com/badge?p=omit.js)](https://packagephobia.com/result?p=omit.js)|^2.0.2|
|open|[![install size](https://packagephobia.com/badge?p=open)](https://packagephobia.com/result?p=open)|^7.0.0|
|ora|[![install size](https://packagephobia.com/badge?p=ora)](https://packagephobia.com/result?p=ora)|^5.0.0|
|p-event|[![install size](https://packagephobia.com/badge?p=p-event)](https://packagephobia.com/result?p=p-event)|^4.2.0|
|p-filter|[![install size](https://packagephobia.com/badge?p=p-filter)](https://packagephobia.com/result?p=p-filter)|^2.1.0|
|p-map|[![install size](https://packagephobia.com/badge?p=p-map)](https://packagephobia.com/result?p=p-map)|^4.0.0|
|p-wait-for|[![install size](https://packagephobia.com/badge?p=p-wait-for)](https://packagephobia.com/result?p=p-wait-for)|^3.0.0|
|parallel-transform|[![install size](https://packagephobia.com/badge?p=parallel-transform)](https://packagephobia.com/result?p=parallel-transform)|^1.2.0|
|parse-github-url|[![install size](https://packagephobia.com/badge?p=parse-github-url)](https://packagephobia.com/result?p=parse-github-url)|^1.0.2|
|parse-gitignore|[![install size](https://packagephobia.com/badge?p=parse-gitignore)](https://packagephobia.com/result?p=parse-gitignore)|^1.0.1|
|path-exists|[![install size](https://packagephobia.com/badge?p=path-exists)](https://packagephobia.com/result?p=path-exists)|^4.0.0|
|path-key|[![install size](https://packagephobia.com/badge?p=path-key)](https://packagephobia.com/result?p=path-key)|^3.1.1|
|path-type|[![install size](https://packagephobia.com/badge?p=path-type)](https://packagephobia.com/result?p=path-type)|^4.0.0|
|prettyjson|[![install size](https://packagephobia.com/badge?p=prettyjson)](https://packagephobia.com/result?p=prettyjson)|^1.2.1|
|pump|[![install size](https://packagephobia.com/badge?p=pump)](https://packagephobia.com/result?p=pump)|^3.0.0|
|raw-body|[![install size](https://packagephobia.com/badge?p=raw-body)](https://packagephobia.com/result?p=raw-body)|^2.4.1|
|resolve|[![install size](https://packagephobia.com/badge?p=resolve)](https://packagephobia.com/result?p=resolve)|^1.12.0|
|semver|[![install size](https://packagephobia.com/badge?p=semver)](https://packagephobia.com/result?p=semver)|^7.3.4|
|source-map-support|[![install size](https://packagephobia.com/badge?p=source-map-support)](https://packagephobia.com/result?p=source-map-support)|^0.5.19|
|static-server|[![install size](https://packagephobia.com/badge?p=static-server)](https://packagephobia.com/result?p=static-server)|^2.2.1|
|strip-ansi-control-characters|[![install size](https://packagephobia.com/badge?p=strip-ansi-control-characters)](https://packagephobia.com/result?p=strip-ansi-control-characters)|^2.0.0|
|tempy|[![install size](https://packagephobia.com/badge?p=tempy)](https://packagephobia.com/result?p=tempy)|^1.0.0|
|through2-filter|[![install size](https://packagephobia.com/badge?p=through2-filter)](https://packagephobia.com/result?p=through2-filter)|^3.0.0|
|through2-map|[![install size](https://packagephobia.com/badge?p=through2-map)](https://packagephobia.com/result?p=through2-map)|^3.0.0|
|to-readable-stream|[![install size](https://packagephobia.com/badge?p=to-readable-stream)](https://packagephobia.com/result?p=to-readable-stream)|^2.1.0|
|toml|[![install size](https://packagephobia.com/badge?p=toml)](https://packagephobia.com/result?p=toml)|^3.0.0|
|update-notifier|[![install size](https://packagephobia.com/badge?p=update-notifier)](https://packagephobia.com/result?p=update-notifier)|^5.0.0|
|uuid|[![install size](https://packagephobia.com/badge?p=uuid)](https://packagephobia.com/result?p=uuid)|^8.0.0|
|wait-port|[![install size](https://packagephobia.com/badge?p=wait-port)](https://packagephobia.com/result?p=wait-port)|^0.2.2|
|winston|[![install size](https://packagephobia.com/badge?p=winston)](https://packagephobia.com/result?p=winston)|^3.2.1|
|write-file-atomic|[![install size](https://packagephobia.com/badge?p=write-file-atomic)](https://packagephobia.com/result?p=write-file-atomic)|^3.0.0|

As Business Competitors, vercel's cli install size is:
| package | install-size  |
| :----- | :------ |
| vercel | [![install size](https://packagephobia.com/badge?p=vercel)](https://packagephobia.com/result?p=vercel) |

vercel cli is only 81mb, and netlify cli is 330mb+.  .....

What a crazy big install-size ! That's why netlify user complained about this issue. https://github.com/netlify/cli/issues/494 [Huge increase in install size]. The core issue is netlify-cli rely on some binary npm package which size is too big. For example, @netlify/routing-local-proxy -> [![install size](https://packagephobia.com/badge?p=@netlify/routing-local-proxy)](https://packagephobia.com/result?p=@netlify/routing-local-proxy) . This binary file size is more than ~50mb. Replace `lodash` with `lodash/fp` or `lodash.xx` is not useful for reduce package install-size, beacuse it's only ~1.3mb.

### Alternative solution
Rewriten netlify-cli in golang, I think these binary netlify private package is also writen in golang, and then pack it as npm package, distribute it via npmjs.org. So netlify user only need install a single platform-specific golang binary file.  I know this is  impossiable, rewriten is a huge work and stop the cli iteration. 

So how about developer experience ? sorry, nobody cares. 
 
cc @erezrokah @ehmicky

**Comment by ehmicky** [comment]
Hi @iChenLei,

Thanks for reporting this issue.

Netlify CLI has lots of features and therefore lots of dependencies. I do agree though that we should strive for a smaller package size. I also agree that 330MB is a lot.

However, the developer experience is mostly only impacted by how long it takes to run `npm install netlify-cli`, which is rather minor IMHO.

Also, this is not an easy fix. As your correctly pointed out with the long table, the total installation size does not come so much from specific dependencies than from the number of dependencies we use. However, most of those dependencies are providing Netlify CLI with essential features.

One potential action item i could see would be to make `@netlify/plugin-edge-handlers` optional, which is something we've been discussing.

Happy to hear what @erezrokah also thinks about this. :thinking:

**Comment by iChenLei** [comment]
Thanks for reply, `@netlify/plugin-edge-handlers` | `@netlify/routing-local-proxy` | `@netlify/zip-it-and-ship-it` | `@netlify/build` ,  I'm curious why these package install-size is so huge. It looks like same with esbuild(writen in golang, and compile different platform-specific binary artifacts ?).

**Comment by iChenLei** [comment]
Anyway, I'm just making a fuss .  Skip it , and continue your great work for netlify-cli. Thanks

**Comment by xmedeko** [comment]
I just need to deploy a small static site (a few kB). If would be nice to have some lite version of netlify-cli for netlify API calls like init, deploy, etc. (i.e. without dev server and others).

**Comment by jtojnar** [comment]
Switching to https://gitlab.com/lepovirta/netlify-deployer made the deploys by CI significantly faster for me.

**Comment by XhmikosR** [comment]
@ehmicky Like I mentioned in #494, this is NOT only an issue of package size.

It's a fundamental security issue: you make all of the people who install your package trust 1300+ packages!

* https://npm.anvaka.com/#/view/2d/netlify-cli

The security chain attacks are real, and the only solution is to reduce the packages you depend on. If you do that, then size and install time will also improve.

Also, the disk churn is serious with each netlify-cli release:

```
C:\Users\xmr\Desktop>npm i -g netlify-cli
npm WARN deprecated flatten@1.0.3: flatten is deprecated in favor of utility frameworks such as lodash.
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated rollup-plugin-inject@3.0.2: This package has been deprecated and is no longer maintained. Please use @rollup/plugin-inject.
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm WARN deprecated node-pre-gyp@0.13.0: Please upgrade to @mapbox/node-pre-gyp: the non-scoped node-pre-gyp package is deprecated and only the @mapbox scoped package will recieve updates in the future

changed 1374 packages, and audited 1375 packages in 25s

108 packages are looking for funding
  run `npm fund` for details

16 vulnerabilities (9 moderate, 7 high)
```

And that's with npm 8.x.

Size is one important factor. Disk churn another. Security is the most important, though.

**Comment by erezrokah** [comment]
Thanks for the additional context @XhmikosR.

I agree the security aspect is more important than the package size.

We've recently dropped `oclif` from the codebase [which helped](https://github.com/netlify/cli/pull/3717#issuecomment-979922158) and are also in the process of converting the codebase to [ES Modules](https://github.com/netlify/cli/pull/3697) so we can upgrade our dependencies to latest versions.

As for the `npm audit` results - unfortunately they are not very helpful as they contain too much noise. It's hard to know which of the reported vulnerabilities are actually exploitable in the context of the CLI.
There's more about that [here](https://news.ycombinator.com/item?id=27761334) and [here](https://overreacted.io/npm-audit-broken-by-design/).

We use a [shrinkwrap file](https://docs.npmjs.com/cli/v8/commands/npm-shrinkwrap), which means once we publish the CLI  you'd always get the same dependencies when you install it using `npm` (with `yarn` you should use your own lock file).
That pins the dependencies users get when installing the CLI, so they don't implicitly get minor/patch updates due to `semver` ranges.

We also highly encourage submitting any security issue you find by following [our policy](https://github.com/netlify/cli/security/policy).

**Comment by XhmikosR** [comment]
I've submitted a few PRs the last couple weeks reducing the deps etc. I've pinpointed the root cause for the biggest size increase in netlify-cli. It comes from multiple typescript versions. See https://github.com/netlify/zip-it-and-ship-it/pull/951 for more info.

This affects quite a few packages that depend on `@netlify/zip-it-and-ship-it`.

**Comment by XhmikosR** [comment]
So, with the above zip-it-and-ship-it change I get these results for netlify-cli:

```
Before: 240 MB (251.943.583 bytes)
After:  188 MB (197.329.508 bytes)
1351 -> 1328 packages
```

I'm sure we can get this even more down.
